### PR TITLE
Enable is_test attr of batch norm and drop out op for test program

### LINKED ISSUE #8372: Need to change the attribute is_test of batch_norm op to true in test_program and inference_program
There is a kind of operators which behaves differently in training and testing phase. They holds a attribute `is_test` to distinguish the type of task, such as [batch_norm_op](https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/fluid/operators/batch_norm_op.cc#L89) and [dropout_op](https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/fluid/operators/dropout_op.cc#L53). We need to set `is_test` to `true` in [test_program](https://github.com/PaddlePaddle/Paddle/blob/develop/python/paddle/v2/fluid/tests/book/test_image_classification.py#L117) and [inference_program](https://github.com/PaddlePaddle/Paddle/blob/develop/python/paddle/v2/fluid/tests/book/test_image_classification.py#L181).

### PR DESCRIPTION
**Pull request by kexinzhao**
fix #8372

### CODE REVIEW

File: paddle/fluid/framework/prune.cc [hunk 1]
```diff
@@ -186,9 +186,7 @@ void Prune(const proto::ProgramDesc& input, proto::ProgramDesc* output) {
   prune_impl(input, output, 0, -1, dependent_vars);
 }
 
-void inference_optimize_impl(const proto::ProgramDesc& input,
-                             proto::ProgramDesc* output, int block_id) {
-  *output = input;
```
Comment (Xreki) [inline paddle/fluid/framework/prune.cc]: output -> inout

File: paddle/fluid/framework/prune.cc [hunk 2]
```diff
@@ -205,7 +203,12 @@ void inference_optimize_impl(const proto::ProgramDesc& input,
```
Comment (Xreki) [inline paddle/fluid/framework/prune.cc]: 这么看，原来的`inference_program`已经有这个修改了啊，那原来是我没注意到。我理解，无论是哪个op，只要有`is_test`这个属性，`inference_program`里面都应该设置成`true`才对，`line 192 - 193`对op type的判断可以去掉？

Comment (kexinzhao) [reply paddle/fluid/framework/prune.cc]: Make sense.  Done.

File: paddle/fluid/framework/prune.cc (hunk 1 above)
Comment (kexinzhao) [reply paddle/fluid/framework/prune.cc]: Done.

File: paddle/fluid/framework/prune.cc (hunk 2 above)
Comment (Xreki) [reply paddle/fluid/framework/prune.cc]: 可以删除`line 30 - 31`

File: python/paddle/fluid/tests/book/test_image_classification.py [hunk 3]
```diff
@@ -115,7 +115,7 @@ def train(net_type, use_cuda, save_dirname, is_local):
     acc = fluid.layers.accuracy(input=predict, label=label)
 
     # Test program 
-    test_program = fluid.default_main_program().clone()
+    test_program = fluid.default_main_program().clone_as_test_program()
```
Comment (Xreki) [inline python/paddle/fluid/tests/book/test_image_classification.py]: 最好是所有的生成`test_program`都改一下？不然用户容易迷惑。

File: paddle/fluid/framework/prune.cc (hunk 2 above)
Comment (kexinzhao) [reply paddle/fluid/framework/prune.cc]: Done

File: python/paddle/fluid/tests/book/test_image_classification.py (hunk 3 above)
Comment (kexinzhao) [reply python/paddle/fluid/tests/book/test_image_classification.py]: 所有三个生成test_program的地方都改了～

File: python/paddle/fluid/framework.py [hunk 4]
```diff
@@ -955,6 +955,14 @@ def clone(self):
...
         return p
 
+    def clone_as_test_program(self):
+        p = Program()
+        p.desc = core.inference_optimize(self.desc)
+        p.blocks = [Block(p, i) for i in xrange(self.desc.num_blocks())]
+        p.sync_with_cpp()
+        p.copy_param_info_from(self)
+        return p
+
```
[the remark this answers is not in the document]
Comment (kexinzhao) [reply python/paddle/fluid/framework.py]: Can't agree more.  Done.

File: python/paddle/fluid/tests/book/test_image_classification.py (hunk 3 above)
Comment (Xreki) [reply python/paddle/fluid/tests/book/test_image_classification.py]: 哦，我原以为所有的示例都有test呢。。。
